waspc/cli/Command/Common.hs

```diff
@@ -25,30 +26,44 @@ import System.Directory
   )
 import qualified System.FilePath as FP
 
-findWaspProjectRoot :: Path Abs (Dir ()) -> Command (Path Abs (Dir WaspProjectDir))
-findWaspProjectRoot currentDir = do
+findWaspProjectRootIO :: Path Abs (Dir ()) -> IO (Maybe (Path Abs (Dir WaspProjectDir)))
+findWaspProjectRootIO currentDir = do
   let absCurrentDirFp = SP.toFilePath currentDir
-  doesCurrentDirExist <- liftIO $ doesPathExist absCurrentDirFp
-  unless doesCurrentDirExist (throwError notFoundError)
-  let dotWaspRootFilePath = absCurrentDirFp FP.</> SP.toFilePath dotWaspRootFileInWaspProjectDir
-  isCurrentDirRoot <- liftIO $ doesFileExist dotWaspRootFilePath
-  if isCurrentDirRoot
-    then return $ SP.castDir currentDir
+  doesCurrentDirExist <- doesPathExist absCurrentDirFp
+  if (not doesCurrentDirExist)
+    then return Nothing
     else do
-      let parentDir = SP.parent currentDir
-      when (parentDir == currentDir) (throwError notFoundError)
-      findWaspProjectRoot parentDir
-  where
-    notFoundError =
-      CommandError
-        ( "Couldn't find wasp project root - make sure"
-            ++ " you are running this command from Wasp project."
-        )
-
-findWaspProjectRootDirFromCwd :: Command (Path Abs (Dir WaspProjectDir))
-findWaspProjectRootDirFromCwd = do
-  absCurrentDir <- liftIO getCurrentDirectory
-  findWaspProjectRoot (fromJust $ SP.parseAbsDir absCurrentDir)
+      let dotWaspRootFilePath = absCurrentDirFp FP.</> SP.toFilePath dotWaspRootFileInWaspProjectDir
+      dotWaspRootFileExists <- doesFileExist dotWaspRootFilePath
+      if dotWaspRootFileExists
+        then return $ Just (SP.castDir currentDir)
+        else do
+          let parentDir = SP.parent currentDir
+          if (parentDir == currentDir)
+            then return Nothing
+            else findWaspProjectRootIO parentDir
+
+findWaspProjectRootCmd :: Path Abs (Dir ()) -> Command (Path Abs (Dir WaspProjectDir))
+findWaspProjectRootCmd currentDir = do
+  maybeWaspProjectRoot <- liftIO (findWaspProjectRootIO currentDir)
+  case maybeWaspProjectRoot of
+    Nothing -> throwError notFoundError
+    Just waspProjectRoot -> return waspProjectRoot
+
+findWaspProjectRootDirFromCwdIO :: IO (Maybe (Path Abs (Dir WaspProjectDir)))
+findWaspProjectRootDirFromCwdIO = do
+  absCurrentDir <- getCurrentDirectory
+  maybeWaspProjectRoot <- findWaspProjectRootIO (fromJust $ SP.parseAbsDir absCurrentDir)
+  case maybeWaspProjectRoot of
+    Nothing -> return Nothing
+    Just waspProjectRoot -> return $ Just (SP.castDir waspProjectRoot)
```

Comment:
You can turn this `case` block into one-liner by using `fmap`! (or `<$>` since it is alias for `fmap`). What enables this is that `Maybe` is Functor, and the fact that you are letting `Nothing` remain `Nothing`.